core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -255,11 +255,18 @@ private MapIterator(int numRecords, Location loc, boolean destructive) {
     }
 
     private void advanceToNextPage() {
+      // SPARK-26265: We will first lock this `MapIterator` and then `TaskMemoryManager` when going
+      // to free a memory page by calling `freePage`. At the same time, it is possibly that another
+      // memory consumer first locks `TaskMemoryManager` and then this `MapIterator` when it
+      // acquires memory and causes spilling on this `MapIterator`. To avoid deadlock here, we keep
+      // reference to the page to free and free it after releasing the lock of `MapIterator`.
```

Comment:
Do we have a deadlock situation for OffHeap situation, too? Does the new test case cover that?

Reply:
Yea, OffHead also suffers from this issue. One change is needed for this test case to cover it. Thanks.